include/pybind11/gil.h

```diff
@@ -120,7 +120,11 @@ class gil_scoped_acquire {
                 pybind11_fail("scoped_acquire::dec_ref(): internal error!");
             }
 #    endif
+            // Make sure that PyThreadState_Clear is not recursively called by finalizers.
+            // See issue #5827
+            ++tstate->gilstate_counter;
```

Comment:
Totally not sure: Is there any point in guarding this with `#if defined(...)` for certain Python versions and `Py_GIL_DISABLED`?